Non-ssl artemis acceptor for RPC connection.

patrickkuo opened the pull request:
* New non-ssl acceptor in artemis server for RPC connection.
* Removed certificate settings in node explore

shamsasari commented on node/src/main/kotlin/net/corda/node/driver/Driver.kt:
```diff
@@ -102,6 +102,7 @@ interface DriverDSLInternalInterface : DriverDSLExposedInterface {
 data class NodeHandle(
         val nodeInfo: NodeInfo,
         val rpc: CordaRPCOps,
+        val rpcAddress: HostAndPort,
```
This is already in `FullNodeConfiguration`. You need to change `rpcClientToNode()` below to use this.

shamsasari commented on node/src/main/kotlin/net/corda/node/driver/Driver.kt:
```diff
@@ -343,6 +344,7 @@ open class DriverDSL(
                            rpcUsers: List<User>, customOverrides: Map<String, Any?>): ListenableFuture<NodeHandle> {
         val messagingAddress = portAllocation.nextHostAndPort()
+        val rpcAddress = portAllocation.nextHostAndPort()
         val apiAddress = portAllocation.nextHostAndPort()
```
rename to webAddress

shamsasari commented on node/src/main/kotlin/net/corda/node/services/config/NodeConfiguration.kt:
```diff
@@ -65,13 +65,14 @@ class FullNodeConfiguration(override val baseDirectory: Path, val config: Config
     // TODO This field is slightly redundant as artemisAddress is sufficient to hold the address of the node's MQ broker.
     // Instead this should be a Boolean indicating whether that broker is an internal one started by the node or an external one
-    val messagingServerAddress: HostAndPort? by config.getOrElse { null }
+    val messagingServerAddress: HostAndPort? by config
```
Is `getOrElse` still used? Can it be removed?

patrickkuo replied:
getOrElse is still used to provide default values in case we don't want nulls for the optional properties.

shamsasari commented on node/src/main/kotlin/net/corda/node/services/messaging/ArtemisMessagingServer.kt:
```diff
@@ -81,7 +81,8 @@ import javax.security.cert.X509Certificate
 class ArtemisMessagingServer(override val config: NodeConfiguration,
-                             val myHostPort: HostAndPort,
+                             val artemisHostPort: HostAndPort,
+                             val rpcHostPort: HostAndPort?,
```
You could just use `config.rpcHostPort`

patrickkuo replied:
rpcHostPort is only accessible from FullNodeConfiguration like artemisAddress...

patrickkuo replied:
rpc's host and port is not accessible from `NodeConfiguration` (same as artemis host and port)...

patrickkuo commented:
Thanks @shamsasari, I have address most issues.
The issue in `ArtemisMessagingServer.kt`(getting rpc address from `NodeConfiguration`) will require moving rpcAdress from `FullNodeConfiguration` to `NodeConfiguration`, which will then inconsistence with existing `artemisAdress`....
Whats your thought?

shamsasari commented:
Yeah sorry ignore that comment, I didn't realised the server wasn't receiving a FullNodeConfiguration. The distinction between NodeConfiguration and FullNodeConfiguration is not clear as some places use the former whilst others use the latter. It requires a clean up, not for this PR.

shamsasari commented:
You've not updated `NodeHandle.rpcClientToNode` to use the rpcAddress. The fact all the tests still pass implies artemisAddress can still be used for RPC. We want to turn that off - only rpcAddress can do RPC and nothing else. Can you also add the relevant tests to `P2PSecurityTest`.

shamsasari commented:
You've not updated `NodeHandle.rpcClientToNode` to use the rpcAddress. The fact all the tests still pass implies artemisAddress can still be used for RPC. We want to turn that off - only rpcAddress can do RPC and nothing else. Can you also add the relevant tests to `P2PSecurityTest`.

patrickkuo commented:
@shamsasari Thanks, will get the unit test done!

Do we want to block off RPC from the artemis address? I was thinking this might be useful when the user want to connect RPC/UI to the server from outside, which they will have to provide valid certificates.

shamsasari commented:
It's easier to reason about the node's security if the P2P port couldn't do RPC. Can we not have SSL optional on the RPC address? @mikehearn may have to chime in here.

mnesbit commented on node/src/main/kotlin/net/corda/node/services/messaging/ArtemisMessagingServer.kt:
```diff
@@ -139,15 +140,22 @@ class ArtemisMessagingServer(override val config: NodeConfiguration,
-        acceptorConfigurations = setOf(tcpTransport(Inbound, "0.0.0.0", myHostPort.port))
+        val acceptors = mutableSetOf(tcpTransport(Inbound, "0.0.0.0", artemisHostPort.port))
+        if (rpcHostPort != null) {
+            acceptors.add(tcpTransport(Inbound, "0.0.0.0", rpcHostPort.port, enableSSL = false))
```
Perhaps we should respect the artemisHostPort.host, rather than using 0.0.0.0. I can see the machine being multihomed with one for users and one for node to node.

patrickkuo replied:
I will do this in a separate PR, will need more testing to make sure changing 0.0.0.0 to the provided host won't break tests and demos...

mnesbit commented on node/src/test/kotlin/net/corda/node/services/messaging/ArtemisMessagingTests.kt:
```diff
@@ -230,8 +231,8 @@ class ArtemisMessagingTests {
-    private fun createMessagingServer(local: HostAndPort = hostAndPort): ArtemisMessagingServer {
-        return ArtemisMessagingServer(config, local, networkMapCache, userService).apply {
+    private fun createMessagingServer(local: HostAndPort = hostAndPort, rpc: HostAndPort = rpcHostAndPort): ArtemisMessagingServer {
+        return ArtemisMessagingServer(config, local,rpc, networkMapCache, userService).apply {
```
Space after comma

patrickkuo commented:
Hi @shamsasari, @mnesbit,
I have addressed most of the issues, I have restricted RPC connection to RPC port only and also added unit test to make sure p2p connection can't connect to RPC port and vice versa.
Regarding changing 0.0.0.0 to provided hostname in the acceptor, i think maybe its better to do that in a different PR.
Thanks.

shamsasari commented on node/src/integration-test/kotlin/net/corda/services/messaging/MQSecurityAsNodeTest.kt:
```diff
@@ -47,4 +51,12 @@ class MQSecurityAsNodeTest : MQSecurityTest() {
     }
+
+    @Test
+    fun `login to a non ssl port`() {
```
login rpc to a non ssl port

shamsasari commented on node/src/integration-test/kotlin/net/corda/services/messaging/MQSecurityAsRPCTest.kt:
```diff
@@ -2,14 +2,35 @@ package net.corda.services.messaging
\ No newline at end of file
+
+    @Test
+    fun `login to a ssl port`() {
```
login rpc to a ssl port

shamsasari commented on node/src/integration-test/kotlin/net/corda/services/messaging/MQSecurityAsRPCTest.kt:
```diff
@@ -2,14 +2,35 @@ package net.corda.services.messaging
+    @Test
+    fun `login to a ssl port`() {
+        val attacker = clientTo(alice.configuration.artemisAddress)
+        Assertions.assertThatExceptionOfType(ActiveMQSecurityException::class.java).isThrownBy {
```
static import assertThatExceptionOfType

shamsasari commented on node/src/integration-test/kotlin/net/corda/services/messaging/MQSecurityAsNodeTest.kt:
```diff
@@ -47,4 +51,12 @@ class MQSecurityAsNodeTest : MQSecurityTest() {
+
+    @Test
+    fun `login to a non ssl port`() {
+        val attacker = clientTo(alice.configuration.rpcAddress!!, null)
```
rename the second param to `ssl` or similar and then refer to it by it's name so it's clear we're not using ssl.

shamsasari commented on node/src/main/kotlin/net/corda/node/services/messaging/ArtemisMessagingServer.kt:
```diff
@@ -480,40 +487,42 @@ class NodeLoginModule : LoginModule {
-        } else {
-            // Otherwise assume they're an RPC user
-            authenticateRpcUser(password, username)
+        val  validatedUser = when (determineUserRole(certificates, username)) {
```
extra space after val

shamsasari commented on node/src/main/kotlin/net/corda/node/services/messaging/ArtemisMessagingServer.kt:
```diff
@@ -480,40 +487,42 @@ class NodeLoginModule : LoginModule {
+            PEER_ROLE -> authenticatePeer(certificates)
+            NODE_ROLE -> authenticateNode(certificates)
+            RPC_ROLE -> authenticateRpcUser(password, username)
+            else -> {
```
Probably don't need the braces

shamsasari commented on test-utils/src/main/kotlin/net/corda/testing/messaging/SimpleMQClient.kt:
```diff
@@ -2,22 +2,22 @@ package net.corda.testing.messaging
 
-    fun start(username: String? = null, password: String? = null) {
-        val tcpTransport = tcpTransport(Outbound(), target.hostText, target.port)
+    fun start(username: String? = null, password: String? = null, enableSSH: Boolean = true) {
```
enableSSL

patrickkuo commented:
Thanks @shamsasari , I have fixed those issues.
Also changed `MQSecurityAsNodeTest.login to a non ssl port` to login as different users instead. The one I had before wasn't right.
